### PR TITLE
initial jetstream API advisories

### PR DESCRIPTION
**Pull request by ripienaar**
Signed-off-by: R.I.Pienaar <rip@devco.net>

### CODE REVIEW

File: surveyor/jetstream_advisory.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+type jsAdvisoryOptions struct {
+	AccountName string `json:"name"`
+	Credentials string `json:"credentials"`
+}
+
+func (o *jsAdvisoryOptions) Validate() error {
+	errs := []string{}
+
+	if o.AccountName == "" {
```
Comment (ColinSullivan1) [inline surveyor/jetstream_advisory.go:44]: Allow $G and old style auth?

Comment (ripienaar) [reply surveyor/jetstream_advisory.go:45]: user/pass is supported , the name here is just for users to know what they are looking at, something for them better than $G

**Comment by ripienaar** [comment]
@ColinSullivan1 minus actual dashboard improvements yet, but I think this is good for reviewing

File: surveyor/jetstream.go [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	}, []string{"account", "stream", "consumer"})
+
+	// Ack Samples
+	jsAckMetricDelay = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_acknowledgement_duration"),
+		Help: "How long an Acknowledged message took to be Acknowledged",
+	}, []string{"account", "stream", "consumer"})
+
+	jsAckMetricDeliveries = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_acknowledgement_deliveries"),
```
Comment (ColinSullivan1) [inline surveyor/jetstream.go:90]: Per prometheus recommendations, should this be `jetstream_acknowedgement_delivery_count`?

File: surveyor/jetstream_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	if err != nil {
+		t.Fatalf("jetstream load error: %s", err)
+	}
+	defer obs.Stop()
+
+	err = obs.Start()
+	if err != nil {
+		t.Fatalf("jetstream failed to start: %s", err)
+	}
+
```
Comment (ColinSullivan1) [inline surveyor/jetstream_test.go:56]: Nitpick...  should probably move the `defer obs.Stop()` above to right here.

File: surveyor/surveyor.go [hunk 4]
```diff
@@ -114,8 +116,8 @@ func connect(opts *Options) (*nats.Conn, error) {
 		nopts = append(nopts, nats.UserInfo(opts.NATSUser, opts.NATSPassword))
 	}
 
-	nopts = append(nopts, nats.DisconnectHandler(func(_ *nats.Conn) {
-		log.Println("Disconnected, will possibly miss replies")
+	nopts = append(nopts, nats.DisconnectErrHandler(func(_ *nats.Conn, err error) {
+		log.Printf("Disconnected, will possibly miss replies: %s", err)
```
Comment (ColinSullivan1) [inline surveyor/surveyor.go:120]: Thanks... :)

**Comment by ColinSullivan1** [comment]
Could you add file format and description in the README?

File: .github/workflows/go.yaml [hunk 5]
```diff
@@ -40,7 +40,6 @@ jobs:
                --enable stylecheck \
                --enable interfacer \
                --enable unconvert \
-               --enable dupl \
```
Comment (ColinSullivan1) [inline .github/workflows/go.yaml:43]: Why remove this?

File: surveyor/jetstream.go (hunk 2 above)
Comment (ripienaar) [reply surveyor/jetstream.go:90]: yup, will fix

File: surveyor/jetstream_test.go (hunk 3 above)
Comment (ripienaar) [reply surveyor/jetstream_test.go:56]: ah, not a Nipick, its just wrong :)

File: .github/workflows/go.yaml (hunk 5 above)
Comment (ripienaar) [reply .github/workflows/go.yaml:43]: startObservations and startJetStreamAdvisories is almost identical and it moans, and yes its almost identical and its a legit gripe, but generic helper would be more complicated and I'd say if we get a 3rd similar its worth extracting into a helper, for 2 its just not worth it

**Comment by ripienaar** [comment]
> Could you add file format and description to the README?

yup, will do towards the end when I have dashboards and know this code works right then I'll do README last

File: .github/workflows/go.yaml (hunk 5 above)
Comment (ColinSullivan1) [reply .github/workflows/go.yaml:43]: +1

**Comment by ColinSullivan1** [comment]
LGTM, feel free to merge when it's ready...

**Comment by ripienaar** [comment]
@ColinSullivan1 so I updated this for some latest things, but worth noting that service latencies have changed format in the jetstream branch so this will not be usable on a old NATS anymore and we cant release this till nats server 2.2.0 - and then thats the required version for latencies.

Not sure what we want to do here - I opened #27 to support the new stuff, shall I merge this one so long? We probably wont have a release till then anyway

**Comment by ColinSullivan1** [comment]
Hmm... there's backward compatibility issues with older servers (2.1.x)...  Imo wait until 2.2 is released to merge, so we remain compatible with the current released version of the server.  This project is new enough (and pre 1.0), I think we can break compatibility for this particular feature.  That being said, if we find we absolutely need to maintain backward compatibility we can check server version and include support for the previous version.  tbh I don't think we'll run into that.

**Comment by ripienaar** [comment]
Hmm I have a lot of changes to make to support the new latency features. It would be awkward to have a bunch of unmerged PRs

**Pull request merged**